jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildResult.java

```diff
@@ -36,4 +37,19 @@ public DescriptorDigest getImageDigest() {
   public DescriptorDigest getImageId() {
     return imageId;
   }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(imageDigest, imageId);
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (this == obj) return true;
+    if (obj == null) return false;
+    if (getClass() != obj.getClass()) return false;
```

Comment:
I think we generally use `instanceof` even though we don't use inheritance, but it does save the extra null check.

Reply:
I meant to look up Bloch previously and he discusses at length in _Effective Java_ how `.equals()` is required to be symmetric and thus `instanceof` should never be used.

Reply:
Filed #1202